File: 70-climbing-stairs/70-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int, memo={}) -> int:
        """
        if n < 0, return 0
        if n == 0, return 1
        can make a decision to take 1 or 2 steps
        depending on step size, decrement n on the next call
        add the return of the recursive calls to a running total
        """
        if n in memo:
          return memo[n]
        if n < 0:
          return 0
        if n == 0:
          return 1
        
        ways = 0
        ways += self.climbStairs(n - 1)
        ways += self.climbStairs(n - 2)
        memo[n] = ways
        return ways

File: 70-climbing-stairs/70-climbing-stairs.java
class Solution {
    private Map<Integer, Integer> memo = new HashMap<>();
    
    public int climbStairs(int n) {
        /*
            the number of ways to get to the n'th step is the sum of the number of ways to get to the previous step and the 2nd previous step.
            dp(i) = dp(i - 1) + dp(i - 2)
            base cases:
                constraint - lowest n can be is 1, so if n == 1, there is only 1 way to climb to the top.
                if n is 2, then there are two ways to climb to the top.
        */
        
        return get_ways(n);
    }
    
    private int get_ways(int n) {
        if (n <= 2) return n;
        if (!memo.containsKey(n)) {
            int num_ways = get_ways(n - 1) + get_ways(n - 2);
            memo.put(n, num_ways);
        }
        
        return memo.get(n);
    }
}